### PR TITLE
Error type revamp distinct type

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Introduce initial set of error type revamp changes.

Fixes #<Issue Number>

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BReadonlyType.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (hasithaa) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BReadonlyType.java:2]: Wrong year

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java [hunk 2]
```diff
@@ -2078,8 +2075,9 @@ private static boolean isEqual(ErrorValue lhsError, ErrorValue rhsError, List<Va
         }
         checkedValues.add(compValuePair);
 
-        return isEqual(lhsError.getReason(), rhsError.getReason(), checkedValues) &&
-                isEqual((MapValueImpl) lhsError.getDetails(), (MapValueImpl) rhsError.getDetails(), checkedValues);
+        return isEqual(lhsError.getMessage(), rhsError.getMessage(), checkedValues) &&
```
Comment (hasithaa) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java:2078]: Need to check cause as well.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/ErrorValue.java [hunk 3]
```diff
@@ -127,18 +139,18 @@ public String toString() {
      *
      * @return reason string
      */
-    @Deprecated
-    public String getReason() {
-        return reason.getValue();
+    // todo: How do we handle this?
+//    public BString getErrorReason() {
```
Comment (hasithaa) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/ErrorValue.java:143]: This is going to replace with type-ids. isn't it?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java [hunk 4]
```diff
@@ -972,9 +974,8 @@ BType readType(int cpI) throws IOException {
                     pkgCpIndex = inputStream.readInt();
                     pkgId = getPackageId(pkgCpIndex);
                     String errorName = getStringCPEntryValue(inputStream);
-                    BType reasonType = readTypeFromCp();
+//                    BType reasonType = readTypeFromCp();
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:977]: Let's remove this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 5]
```diff
@@ -481,14 +480,6 @@ public void visit(BLangRecordTypeNode recordTypeNode) {
...
     public void visit(BLangErrorType errorType) {
-        BType reasonType = getReasonType(errorType);
-
-        if (!types.isAssignable(reasonType, symTable.stringType)) {
-            dlog.error(errorType.reasonType.pos, DiagnosticCode.INVALID_ERROR_REASON_TYPE, reasonType);
-        } else if (errorType.reasonType != null) {
-            validateModuleQualifiedReasons(errorType.reasonType.pos, reasonType);
-        }
-
         if (errorType.detailType == null) {
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:483]: This is no longer valid, isn't it?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java (hunk 2 above)
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java:2078]: Line:2080 does that

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 5 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:483]: We can still have something like `type F error`
https://ballerina.io/ballerina-spec/spec.html#section_5.5.1

**Pull request merged**